File: GO/Structures/StackLinear.go
package Structures

type StackLinear struct {
    data     []int
    top      int
    capacity int
}

func NewStackLinear(capacity int) *StackLinear {
    return &StackLinear{
        data:     make([]int, capacity),
        top:      -1,
        capacity: capacity,
    }
}

func (s *StackLinear) Push(value int) {
    if s.top < s.capacity-1 {
        s.top++
        s.data[s.top] = value
    }
}

func (s *StackLinear) Pop() (int, bool) {
    if s.top == -1 {
        return 0, false
    }
    
    value := s.data[s.top]
    s.top--
    return value, true
}

func (s *StackLinear) ToArray() []int {
    result := make([]int, s.top+1)
    copy(result, s.data[:s.top+1])
    return result
}

func (s *StackLinear) FromArray(arr []int) {
    s.top = -1
    for _, v := range arr {
        s.Push(v)
    }
}

func (s *StackLinear) Size() int {
    return s.top + 1
}

File: PYTHON/structures/stack_linear.py
# structures/stack_linear.py
from typing import List, Tuple
import sys
sys.path.append('..')
from utils.data_structure import DataStructure


class StackLinear(DataStructure):
    """Pilha com alocação linear (array fixo)"""
    
    def __init__(self, capacity: int):
        self.data = [0] * capacity
        self.top = -1
        self.capacity = capacity
    
    def push(self, value: int) -> None:
        if self.top < self.capacity - 1:
            self.top += 1
            self.data[self.top] = value
    
    def pop(self) -> Tuple[int, bool]:
        if self.top == -1:
            return 0, False
        
        value = self.data[self.top]
        self.top -= 1
        return value, True
    
    def to_array(self) -> List[int]:
        return self.data[:self.top + 1].copy()
    
    def from_array(self, arr: List[int]) -> None:
        self.top = -1
        for v in arr:
            self.push(v)
    
    def size(self) -> int:
        return self.top + 1